### PR TITLE
Implement resizing for the file browser

### PR DESCRIPTION
**Pull request by krassowski**
## References

- Closes https://github.com/jupyterlab/jupyterlab/issues/3875
- Fixes https://github.com/jupyterlab/jupyterlab/issues/16239

## Code changes

- after initial load the fixed-width layout is used in favour of the flexbox which is now only used to load initial widths and for backward compatibility
- the filename icon and label are now grouped together into a single node which allows to set a column size on the file name column
- `FileBrowser` and `DirListing` widgets now accept optional `state?: IStateDB` option which allows the listing to restore the size of columns; this follows the pattern already used in the `FileBrowserModel`
- `DirListing.IRenderer` gets a new optional `updateItemSize?(...): void` method which is a fast path to refresh the widths in the listing without updating other elements
- The default implementation of `DirListing.IRenderer.updateItemModified()` will now check if there is any work to be done to avoid costly `format()` calls
- All hidden columns are now detached from the DOM when hidden, which saves time on rendering (which happens often when resizing); previously checkboxes were detached but "Last Modified" and "File Size" were just hidden using CSS

## User-facing changes

- The file browser render and refreshes faster
- Columns can be resized
resize-demo

## Backwards-incompatible changes

None

### CODE REVIEW

**Comment by krassowski** [comment]
Because I had to rework (=improve) the caret positioning, the caret is no longer inflating the height of the column headers. This does not change the height as seen when checkboxes are enabled (because the checkbox in the header is inflating the height too) but it reduces the height otherwise:

| Before | Now |
|---|--|
| image | image |

I personally don't mind having a little bit more space for files, but if there are strong opinions I am happy to reinstate the extra pixel or two - just let me know.

File: packages/filebrowser/src/listing.ts [hunk 1]
```diff
@@ -2722,19 +3077,23 @@ export namespace DirListing {
         node.removeChild(checkboxWrapper);
       } else if (showFileCheckboxes && !checkboxWrapper) {
         const checkboxWrapper = this.createCheckboxWrapperNode();
-        node.insertBefore(checkboxWrapper, iconContainer);
+        nameColumn.insertAdjacentElement('beforebegin', checkboxWrapper);
```
Comment (afshin) [inline packages/filebrowser/src/listing.ts:3080]: Switching to `insertAdjacentElement` is nice.

File: packages/filebrowser/src/listing.ts [hunk 2]
```diff
@@ -2888,6 +3278,34 @@ export namespace DirListing {
       return dragImage;
     }
 
+    /**
+     * Factories for individual parts of the item.
+     */
+    protected _itemFactories = {
```
Comment (afshin) [inline packages/filebrowser/src/listing.ts:3284]:
```suggestion
    protected itemFactories = {
```

(convention for `protected` accessors)

**Review by afshin — APPROVED** [review]
Thank you, @krassowski! I made one minor comment.

**Pull request merged**